test/unit/connection_string.test.ts

```diff
@@ -89,6 +94,33 @@ describe('Connection String', function () {
     expect(options.credentials.source).to.equal('$external');
   });
 
+  it('should retain user-specified authSource with no credentials for non-srv URI', function () {
+    const mockAuthSource = 'thisShouldBeAuthSource';
+    const options = parseOptions(`mongodb://localhost/?authSource=${mockAuthSource}`);
+
+    expect(options).to.have.nested.property('credentials.username', '');
+    expect(options).to.have.nested.property('credentials.mechanism', 'DEFAULT');
+    expect(options).to.have.nested.property('credentials.source', mockAuthSource);
+  });
+
+  it('should omit credentials if the only auth related option is authSource', async () => {
+    const client = new MongoClient('mongodb://localhost:123/?authSource=someDb', {
```

Comment:
Why is this test connecting to a client while the other test above is using `parseOptions`? It's kind of confusing for a unit test, since, in general, if we are testing different functions then they shouldn't all be under the "Connection string" bucket, each helper should have its own describe block (like `resolveSrvRecordAsync` does).

For this specific test case, it's not clear why we have to run it through connect before checking the options.

Reply:
The error we're looking to **not** see is `new MongoInvalidArgumentError('No AuthProvider for ${credentials.mechanism} defined.')` in `prepareHandshakeDocument` and/or `performInitialHandshake`. Neither function is exported currently but if I did export them becuase of the inlined callbacks I think I would need to mock quite a bit of internals to get down to that layer. My thinking is I can lean on server selection failing for th unit tests to assert we at least don't get an error related to auth. I'm happy to comment something to this effect and move the test to a better spot, thoughts?

Reply:
I think it's fine to do what you are saying with the comments, no need for convoluted mocks, but you could even kind of document the intention in the assertions and the test title by asserting that the thrown error is not a MongoAPIError derivative (and we can also make the test title reflect that expectation); obligatory suggestion to split into 2 separate test cases - one that it should not throw, and the other about what the actual result is expected to be

Reply:
The success case requires me to have an actual server to connect to, I think it could be done in an int test but I think its overkill for this case. Additionally because this is covered by the uri-options spec tests, so we'll have a more complete coverage when those are fixed up.